website/blog/modules/ROOT/pages/15-android-build-flow.adoc

```diff
@@ -0,0 +1,223 @@
+  "App Proguard rules" -> "Proguard rules"
+
+  // --- Library inputs
+  subgraph cluster_aar {
+    label="Dependency (AARs)"
+    rankdir=TB
+    node [shape=box width=0 height=0 style=filled fillcolor=white]
+
+    "AAR Files" -> "Unpack AARs"
+    "Unpack AARs" -> "AAR classes.jar"
+    "Unpack AARs" -> "AAR res/"
+    "Unpack AARs" -> "AAR AndroidManifest.xml"
+    "Unpack AARs" -> "AAR proguard.txt"
+    "Unpack AARs" -> "AAR native .so (optional)"
+  }
+
+  // --- Resource/link phase
+  "AAR res/" -> "Compile Lib Resources (aapt2)"
+  "Compile App Resources (aapt2)" -> "Linked Resources"
+  "Compile Lib Resources (aapt2)" -> "Linked Resources"
+  "AAR AndroidManifest.xml" -> "Manifest Merging"
+  "Manifest Merging" -> "Linked Resources"
+
+  // --- Classes & DEX
+  "Compile (Java/Kotlin)" -> "Compiled Classes"
+  "AAR classes.jar" -> "Compile Classpath"
+  "AAR classes.jar" -> "DEX (d8/r8)"
+  "Compile Classpath" -> "Compile (Java/Kotlin)"
+  "Linked Resources" -> "Package APK"
+  "Compiled Classes" -> "DEX (d8/r8)"
+  "DEX (d8/r8)" -> "Package APK"
+
+  // --- Proguard / main-dex rules
+  "AAR proguard.txt" -> "Proguard rules"
+  "Linked Resources" -> "Proguard rules"
+  "Proguard rules" -> "DEX (d8/r8)"
+
+  // --- Native libs & META-INF (optional)
+  "AAR native .so (optional)" -> "Package APK"
+
+  // --- Final steps
+  "Package APK" -> "Code Shrinking (r8/Proguard)"
+  "Code Shrinking (r8/Proguard)" -> "Sign APK"
+  "Sign APK" -> "Install to Emulator"
+  "Install to Emulator" -> "Run/Test via ADB"
+}
+....
+
+The diagram above still doesn’t tell the whole story! It shows a typical build flow for an everyday Android app, but there are more features to consider:
+
+- Hilt/Dagger code generation (annotation processing)
+- Jetpack Compose code generation (Kotlin compiler plugin)
+- Instrumented tests (separate APK, own resources, manifests, dependencies)
+- Native code (NDK builds, CMake integration)
+
+We cover a lot of these architecture styles in various Android examples, based on xref:mill::android/java.adoc[Java], xref:mill::android/kotlin.adoc[Kotlin] and third party integration examples covering xref:mill::android/compose-samples.adoc[Android Compose], xref:mill::android/android-native-example.adoc[Android Native] and xref:mill::android/hilt-sample.adoc[Dependency Injection with Hilt].
+
+Endless tunnel sample app
+image:AndroidEndlessTunnelExample.png[An Android app built with Mill using native code, showing a 3D tunnel effect.]
+
+== Try it out
+
+Mill’s Android support is still young, but it already covers the full build pipeline: resource compilation, manifest merging, packaging, signing, running, and even testing on emulators.
+
+What makes this different from other build tools is control: every build step is a visible Mill task, easy to run on its own, inspect, or override. That means you can debug problems faster, adapt the pipeline to your project’s needs, and extend it without fighting opaque built-in or plugin logic.
```

Comment:
The obvious question to me here is: is this not true of Gradle? Maybe it's not, but you have to show that in more detail for me to understand why Mill is cool